### ISSUE TITLE
[RFC] Math library and physics engine

### ISSUE THREAD

**Issue by vitvakatu** [amethyst/amethyst#205]
# Summary
1. Prefer using `cgmath` in every place inside engine instead of raw types.
2. Implement physics engine integrated with `specs` and `cgmath`
# Motivation
Using specific math library will allow us to focus our forces on developing instead of trying to support bunch of different libraries. Avoiding of using raw types will improve our API and will make it more user-friendly. Still, users will have opportunity to use any library they want, but we should use only one approach inside engine.
Own physics engine based on `specs` will be as easy as possible to use with other parts of engine. Besides it will allow us to use some interesting things like animation system integrated with physics engine.
# Detailed design
1. Export some useful types from `cgmath` to Amethyst's namespace and create short and readable aliases for them. (For example, `Vec3f` for `cgmath::Vector3<f32>`)
2. Use `cgmath`'s types in every place where it is possible. Avoid using raw types and own implementations. One example - `Projection` enum.
3. Separate paragraph, because it's very important. Replace arrays in `InnerTransform` with `Vector`s from `cgmath`. Implement `Into` or `Deref` traits to use it in old functions. It will give us user-friendly API (`local.translation.x` instead of ugly `local.translation[0]`)
4. There is no strict plan for implementing physics engine. Anyway, me and @torkleyy have some ideas and already started working on it. Please, write down you proposals and comments.

# Drawbacks
First of all, a lot of work to do. Probably we could broke backward compatibility, but it's not that case when it's very important.
Other drawbacks depends mostly on specific implementation and connected with physics engine.

# Unresolved questions
A lot of questions about physics engine. Starting with level of isolation from other parts and ending with possible connections with animation system.

**Comment by msiglreith** [comment]
> Export some useful types from cgmath to Amethyst's namespace and create short and readable aliases for them. (For example, Vec3f for cgmath::Vector3<f32>)

👍, could also re-export the whole library. I think this could also prevent some version collision.

> Replace arrays in InnerTransform with Vectors from cgmath. Implement Into or Deref traits to use it in old functions. It will give us user-friendly API (local.translation.x instead of ugly local.translation[0])

Not totally sure about this point. It would be nice to keep the current `[f32; 3]` representation for public components for better compatiblity with user written systems. I don't see that much advantage for replacing them with cgmath vector type.


Regarding physics engine:
We should split the physics part into two libraries for collision and actual physics. Giving user the possiblities to re-use exisiting collision methods for own physics engines (e.g 2d games) and trigger systems, ray-casts, etc. for interaction methods or other gameplay elements.
Animation system could be realized by handling them as kinematic objects, which are only moved by the animation but interact (one-way only) with other physics objects.
Overall, there a bunch of different approaches for implementing physics in real-time. Not sure about the current state of art but most implementation are based on constraint modelling coupled with an iterative solver afaik. (Recent trend moves toward continuous collision solvers but they are even more complex to handle..)

Thanks for starting this RFC, excited to see other proposals and suggestions!

**Comment by vitvakatu** [comment]
@msiglreith 
> Not totally sure about this point. It would be nice to keep the current [f32; 3] representation for public components for better compatiblity with user written systems. I don't see that much advantage for replacing them with cgmath vector type.
Imagine this situation: you need to change player position according to his current velocity.
For now, you should explicitly write it for all coordinates:
```
player.translation[0] += player.velocity.x; 
player.translation[1] += player.velocity.y;
player.translation[2] += player.velocity.z;
```
With `Vector` all you need is:
```
player.translation += player.velocity;
```
Less code, less error prone because of copy-pasting.
I'm still making an accent at this proposal particularly.

**Comment by kvark** [comment]
@vitvakatu 
I agree with the RFC, with just a few notes:

> Implement physics engine integrated with specs and cgmath

I don't think the physics engine needs to know about `specs`, or even about `amethyst`. It needs to be something usable, composable, and designed with multi-threading in mind. Similarly how `gfx-rs` doesn't know  anything about `specs` but plays well with it.

from @msiglreith 
> We should split the physics part into two libraries for collision and actual physics.

Right. For collision, we can make collision-rs more complete.

**Comment by vitvakatu** [comment]
@kvark 
> I don't think the physics engine needs to know about `specs`, or even about `amethyst`. It needs to be something usable, composable, and designed with multi-threading in mind.

I'm not sure about how possible it is. Could we create physics engine, that separated from `amethyst` and, at the same time, usable? IMO, we'll need to create some kind of wrapper around physics engine to allow using it in user's project without pain. Maybe it is redundant? If somebody want to use physics engine without `amethyst`, there is `nphysics`. 
Well, it's difficult question, and I request others opinions.
> we can make collision-rs more complete.

It makes sense. Thank you for resurrecting `collision-rs`.

**Comment by torkleyy** [comment]
I agree with most points of this RFC.

@vitvakatu 

I also think that we want to have a physics engine which isn't even aware of amethyst or specs. We may create some library `foo_physics` and the provide a nice bridge like `amethyst_foo_physics`.

I'm also in favor of extending `collision-rs`. I just looked at `nphysics` and I don't think we could integrate it into `amethyst` nicely. So the difference between our physics engine and `nphysics` would be the structure and probably also that it contains animation logic if we want to build animation into the physics engine.

**Comment by minecrawler** [comment]
What do you think about having a way to share certain data and coordination between the renderer and the physics engine? Modern APIs for graphics and compute (for example Vulkan) offer ways to do both and share data or even use different devices for different tasks. That would, however require a coordination between both parts, which would have to be done inside Amethyst.

The clear benefit is, that a computer with more than one GPU (iGPU or multiple dedicated cards) can intelligently distribute tasks and achieve a speedup vs. completely separated implementations, which might end up using the same GPU, even though another one would be available (which is idle).

**Comment by kvark** [comment]
@minecrawler it's definitely something to keep on our radar, however I'm afraid that focusing on that feature would blow up the scope of the problem. There are no good semi-complete engines in Rust yet, so we better have at least something, even if it's designed for yesterdays hardware.

**Comment by nchashch** [comment]
@torkleyy The idea was not to embed animation into physics engine.

TL;DR: Physics and animation systems should share the same mechanism for feeding time varying data to transform system.

Suppose we have an object with one coordinate `x` (maybe it is sliding on a straight rail) in an `amethyst` game. In order to animate it we would have to compute `x = x(t)` every frame. Notice that it doesn't matter how we get `x(t)`. `x(t)` can be sampled from interpolated hand crafted animation data, from precomputed physics data, from real time physics data, or say from player's x mouse coordinate.

Job of the physics engine then would be to produce "`x(t)`" functions according to some physical laws. For example if you wanted to model a ball flying on a parabolic trajectory, your physics engine would have to produce a `y(t)` and an `x(t)` sampler for vertical and horizontal coordinates. In this case it would make sense to just recompute analytic formulas (`x(t) = x_0 + v_x0*t`, `y(t) = y_0 + v_y0*t + a*t^2/2`). For something that can't be solved analytically you might want to do numeric integration.

It also would be nice to have some mechanism to "replace" samplers. So for example when the ball hits the ground, the initial parabolic trajectory sampler for ball moving with energy `E0` could be replaced with a new parabolic trajectory sampler say for energy `0.7*E0`, so after impact the ball would bounce back and lose 30% of it's kinetic energy. Producing a new sampler and figuring out when to replace samplers would require collision detection.

**Comment by cjhowe** [comment]
I think this is quite a lot of work. The presumption of this RFC seems to be that nphysics wouldn't integrate well with animation. Are there specific examples of why this wouldn't work? Is there maybe some way we can work with the nphysics developer to improve this situation?

**Comment by nchashch** [comment]
@cjhowe I guess suggestions for animation/physics using the same mechanism to move things in the world would probably apply the same for both nphysics/non nphysics based solution.

**Comment by cjhowe** [comment]
I tried to fix `nphysics` so that it could be `Send` and `Sync`, but I don't think that's going to happen. The problem is that the `nphysics` world stores a pointer to itself (and uses an `Rc` to get around this), and that pointer is actually stored in a structure from another crate (`ncollide`). I think one solution would be to change the data flow so that `ncollide` handlers are provided with the `nphysics` world, but the obvious problem with that is that it'd add a circular dependency of `ncollide` on `nphysics`, which I think is against the goal of those crates. Additionally, you could use an `Arc`, but as mentioned in #144, that might not be ideal for performance reasons. So I think maybe our own physics engine may be the way forward, although I think we should keep collision detection and physics in the same crate to avoid this issue.

**Comment by MrMonotone** [comment]
I do not really like the duplicating effort however, it seems like it needs to be done. It might not mean much but the lack of a nice way to do physics and collision is what is stopping me from using amethyst.

**Comment by kvark** [comment]
@vitvakatu is this RFC still active?
There is an existing answer (#236) precisely addressing the motivation:

> Using specific math library will allow us to focus our forces on developing instead of trying to support bunch of different libraries. Avoiding of using raw types will improve our API and will make it more user-friendly. Still, users will have opportunity to use any library they want, but we should use only one approach inside engine.

**Comment by vitvakatu** [comment]
The state of this RFC is unclear for me. I'm not working on Amethyst anymore, dev team can close this issue if you don't want to continue discussion here.

FYI, I didn't change my opinion on every point stated here.

**Comment by Xaeroxe** [comment]
I generally like this proposal as I'm fond of cgmath myself, however the future of physics for Amethyst is unclear.  Right now we have two major physics libraries that are being written or rewritten for compatibility with Amethyst, one uses cgmath, the other uses nalgebra.  I don't believe we should pick a math library until we also have a strong physics engine supporting said library.  That being said I feel nalgebra is rather obtuse about relatively simple things so I'd prefer to use the much simpler cgmath.

**Comment by minecrawler** [comment]
both, cgmath and nalgebra, implement Mint! We don't have to even care about which to use, as long as we just use mint interfaces, which makes stuff very convenient. Afaik, it's the reason why Mint was created. So we can just decide to use cgmath in Amethyst, use Mint in APIs and use the nalgebra physics engine, if we wanted to do so.

**Comment by vitvakatu** [comment]
@minecrawler, yes, you're right, it was the very reason

**Comment by Aceeri** [comment]
I'm going to say we should go forward with using mint.
